File: app/lib/ir/hypothesis_gen_for_ir_from_bin.py
# hypothesis_gen_for_ir_from_bin.py

import json
import struct
import math
import sys
from pathlib import Path


USIG_MAGIC_HINTS = [
    b'"waveformLength"',
    b'"waveformEncoding"',
    b'"schemaVersion"',
    b'USIG'
]


# ------------------------------------------------------------
# IO
# ------------------------------------------------------------

def read_file(path):
    return Path(path).read_bytes()


def safe_decode_ascii(data):
    return data.decode("utf-8", errors="ignore")


# ------------------------------------------------------------
# USIG detection
# ------------------------------------------------------------

def detect_usig_ir(data):

    text = safe_decode_ascii(data[:4096])

    hits = sum(
        1
        for x in USIG_MAGIC_HINTS
        if x.decode(errors="ignore") in text
    )

    return hits >= 2, (1.0 if hits >= 2 else 0)


# ------------------------------------------------------------
# JSON extraction
# ------------------------------------------------------------

def extract_json_region(data):

    tokens = [
        b'{"headers"',
        b'{"schemaVersion"',
        b'{"metadata"',
        b'{"waveformLength"',
        b'{"cacheKey"'
    ]

    starts = []

    for token in tokens:
        pos = data.find(token)
        if pos >= 0:
            starts.append(pos)

    for start in sorted(starts):

        depth = 0
        in_string = False
        escape = False

        for i in range(start, len(data)):

            c = data[i]

            if in_string:

                if escape:
                    escape = False

                elif c == 92:
                    escape = True

                elif c == 34:
                    in_string = False

            else:

                if c == 34:
                    in_string = True

                elif c == 123:
                    depth += 1

                elif c == 125:

                    depth -= 1

                    if depth == 0:

                        blob = data[start:i+1]

                        try:
                            return json.loads(
                                blob.decode("utf-8")
                            )

                        except Exception:
                            break

    return None



def flatten_json(obj, prefix=""):

    out = []

    if isinstance(obj, dict):

        for k, v in obj.items():

            path = (
                f"{prefix}.{k}"
                if prefix
                else k
            )

            if isinstance(v, (dict, list)):
                out.extend(
                    flatten_json(v, path)
                )

            else:

                out.append({
                    "path": path,
                    "value": v,
                    "source": "embedded_json"
                })

    return out



# ------------------------------------------------------------
# waveform definition normalization
# ------------------------------------------------------------

def normalize_encoding(value):

    if not value:
        return None

    v = str(value).lower()

    if "float32" in v:
        return "float32"

    if "int16" in v:
        return "int16"

    return v



def build_waveform_definition(
        fields,
        reconstruction=None,
        candidate=None):

    wf = {}

    # --------------------------------------------------------
    # explicit fields first
    # --------------------------------------------------------

    lookup = {}

    for f in fields:

        lookup[f["path"].lower()] = f["value"]


    def find(names):

        for k, v in lookup.items():

            for name in names:

                if k.endswith(name.lower()):
                    return v

        return None


    wf["encoding"] = normalize_encoding(
        find([
            "waveformencoding",
            "waveformtype",
            "encoding"
        ])
    )


    wf["endianness"] = find([
        "endianness"
    ])


    wf["bytes_per_sample"] = find([
        "waveformbytesperelement",
        "bytespersample"
    ])


    wf["samples"] = find([
        "waveformlength",
        "samples"
    ])


    wf["offset"] = 0


    wf["scale"] = find([
        "scale"
    ])


    wf["offset_value"] = find([
        "offset_value"
    ])


    # --------------------------------------------------------
    # reconstruction fallback
    # --------------------------------------------------------

    if reconstruction:

        if wf["encoding"] is None:
            wf["encoding"] = normalize_encoding(
                reconstruction.get("encoding")
            )

        if wf["samples"] is None:
            wf["samples"] = reconstruction.get(
                "length"
            )

        if wf["bytes_per_sample"] is None:

            bl = reconstruction.get(
                "byte_length"
            )

            ln = reconstruction.get(
                "length"
            )

            if bl and ln:
                wf["bytes_per_sample"] = (
                    bl // ln
                )


    # --------------------------------------------------------
    # heuristic fallback
    # --------------------------------------------------------

    if candidate:

        for k in [
            "encoding",
            "endianness"
        ]:

            if wf.get(k) is None:
                wf[k] = candidate.get(k)


        if wf["samples"] is None:
            wf["samples"] = candidate.get(
                "length"
            )

        if wf["offset"] == 0:
            wf["offset"] = candidate.get(
                "offset",
                0
            )


    # --------------------------------------------------------
    # defaults
    # --------------------------------------------------------

    if wf["encoding"] is None:
        wf["encoding"] = "float32"


    if wf["endianness"] is None:
        wf["endianness"] = "little"


    if wf["bytes_per_sample"] is None:

        if wf["encoding"] == "float32":
            wf["bytes_per_sample"] = 4

        elif wf["encoding"] == "int16":
            wf["bytes_per_sample"] = 2


    if wf["scale"] is None:
        wf["scale"] = 1.0


    if wf["offset_value"] is None:
        wf["offset_value"] = 0.0


    return wf



# ------------------------------------------------------------
# USIG parser
# ------------------------------------------------------------

def parse_usig(data):

    meta = extract_json_region(data)

    result = {
        "container": {
            "type": "USIG_IR",
            "confidence": 1.0
        },
        "metadata_fields": [],
        "reconstruction": {}
    }


    if not meta:

        result["container"]["confidence"] = 0.5
        return result


    fields = flatten_json(meta)

    result["metadata_fields"] = fields


    length = None
    byte_length = None
    encoding = None


    for f in fields:

        p = f["path"]

        if p.endswith("waveformLength"):
            length = f["value"]

        if p.endswith("waveformByteLength"):
            byte_length = f["value"]

        if p.endswith("waveformEncoding"):
            encoding = f["value"]


    if length and byte_length:

        result["reconstruction"] = {
            "length": length,
            "byte_length": byte_length,
            "encoding": encoding or "Float32Array"
        }


    return result



# ------------------------------------------------------------
# heuristic detector
# ------------------------------------------------------------

def float_stats(data, offset):

    values = []

    count = (
        len(data)-offset
    ) // 4


    if count < 64:
        return None


    for i in range(count):

        try:

            v = struct.unpack_from(
                "<f",
                data,
                offset+i*4
            )[0]

            if math.isfinite(v):
                values.append(v)

        except:
            break


    if len(values)<64:
        return None


    mn=min(values)
    mx=max(values)
    mean=sum(values)/len(values)


    if abs(mx)>1e10 or abs(mn)>1e10:
        return None


    return {
        "count":len(values),
        "min":mn,
        "max":mx,
        "mean":mean
    }



def find_waveform_candidate(data):

    best=None


    for offset in range(
        0,
        min(len(data),8192),
        4
    ):

        stats=float_stats(
            data,
            offset
        )

        if not stats:
            continue


        confidence=0.5


        if (
            stats["min"]>-10 and
            stats["max"]<10
        ):
            confidence+=0.4


        if abs(stats["mean"])<1:
            confidence+=0.099


        candidate={
            "name":"waveform",
            "offset":offset,
            "encoding":"float32",
            "endianness":"little",
            "length":stats["count"],
            "statistics":stats,
            "confidence":round(
                min(confidence,0.999),
                3
            )
        }


        if (
            best is None or
            candidate["confidence"] >
            best["confidence"]
        ):
            best=candidate


    return best



# ------------------------------------------------------------
# main generator
# ------------------------------------------------------------

def generate_hypothesis(filename):

    data=read_file(filename)


    result={
        "file":Path(filename).name,
        "size_bytes":len(data),
        "container":{},
        "metadata_fields":[],
        "reconstruction_fields":{},
        "arrays":[]
    }


    is_usig,_=detect_usig_ir(data)


    if is_usig:

        parsed=parse_usig(data)

        result.update(parsed)

        result["waveform_definition"] = (
            build_waveform_definition(
                result["metadata_fields"],
                result.get("reconstruction")
            )
        )


        result["decision"]={
            "can_create_ir":True,
            "mode":"direct_waveform"
        }

        return result



    result["container"]={
        "type":"unknown_binary",
        "confidence":0
    }


    candidate=find_waveform_candidate(data)


    if candidate:

        result["arrays"]=[
            candidate
        ]

        result["waveform_definition"]=(
            build_waveform_definition(
                [],
                candidate=candidate
            )
        )

        result["decision"]={
            "can_create_ir":True,
            "mode":"heuristic_waveform"
        }


    else:

        result["decision"]={
            "can_create_ir":False,
            "mode":"unknown"
        }


    return result



# ------------------------------------------------------------
# CLI
# ------------------------------------------------------------

if __name__=="__main__":
    for str_input_bin_file_path in ["/media/gnew/Mech/results/usig_unitesting/wavebin_examples/single.bin",
                                    r"/media/gnew/Mech/results/usig_unitesting/cannon_conversion_files/sine_fs2p25ghz_tonemode~single_fftlength8192_numaveraging4_numberofcores8_ticorrections~ogp_woverrides.bin"]:
        input_path = str_input_bin_file_path

        if not input_path:
            print("No input file provided")
            raise SystemExit(1)

        try:
            output = generate_hypothesis(input_path)

            print(
                json.dumps(
                    output,
                    indent=2
                )
            )

        except FileNotFoundError:
            print(
                f"File not found: {input_path}"
            )
            raise SystemExit(1)

        except Exception as e:
            print(
                f"Processing failed: {e}"
            )
            raise

File: app/lib/ir/hypothesis_gen_for_ir_from_bin.js
#!/usr/bin/env node

// hypothesis_gen_for_ir_from_bin.js
//
// USIG IR Binary Hypothesis Generator
//
// Purpose:
// Analyze arbitrary binary payloads and generate
// reconstruction instructions suitable for ingestion
// into USIG IR.
//
// Detection paths:
//
// 1. Embedded metadata reconstruction
// 2. Heuristic waveform reconstruction
// 3. Reject unknown layouts
//
// This module intentionally does not assume
// a binary file format.
import path from "node:path";

// ------------------------------------------------------------
// Constants
// ------------------------------------------------------------

const METADATA_HINTS = [
    '"waveformLength"',
    '"waveformEncoding"',
    '"schemaVersion"',
    '"waveformBytesPerElement"',
    '"metadata"'
];

const JSON_START_TOKENS = [
    '{'
];

// ------------------------------------------------------------
// IO
// ------------------------------------------------------------

function safeDecodeUtf8(buffer) {
    return buffer.toString("utf8");
}

// ------------------------------------------------------------
// Metadata detection
// ------------------------------------------------------------

function detectEmbeddedMetadata(data) {
    const text =
        safeDecodeUtf8(
            data.subarray(
                0,
                Math.min(
                    data.length,
                    4096
                )
            )
        );

    let hits = 0;

    for (const hint of METADATA_HINTS) {
        if (text.includes(hint)) {
            hits++;
        }
    }

    return {
        detected:
            text.includes('"waveformLength"') &&
            text.includes('"waveformBytesPerElement"'),
        confidence:
            hits >= 2
                ? 1.0
                : 0
    };
}

// ------------------------------------------------------------
// JSON extraction from binary payload
// ------------------------------------------------------------

function extractJsonRegion(data) {

    const text =
        data.toString(
            "utf8",
            0,
            Math.min(data.length, 65536)
        );

    const starts = [];

    for (const token of JSON_START_TOKENS) {
        const index =
            text.indexOf(token);
        if (index >= 0) {
            starts.push(index);
        }
    }

    for (
        const start of starts.sort(
            (a,b)=>a-b
        )
    ) {
        let depth = 0;
        let inString = false;
        let escape = false;

        for (
            let i=start;
            i<text.length;
            i++
        ) {

            const c = text[i];

            if (inString) {
                if (escape) {
                    escape = false;
                }
                else if (c === "\\") {
                    escape = true;
                }
                else if (c === '"') {
                    inString = false;
                }
            }
            else {
                if (c === '"') {
                    inString = true;
                }
                else if (c === "{") {
                    depth++;
                }
                else if (c === "}") {
                    depth--;

                    if (depth === 0) {

                        const candidate =
                            text.substring(
                                start,
                                i + 1
                            );

                        try {

                            return JSON.parse(
                                candidate
                            );

                        }

                        catch(e) {

                            break;

                        }

                    }

                }

            }

        }

    }

    return null;

}

// ------------------------------------------------------------
// Flatten metadata
// ------------------------------------------------------------

function flattenJson(
    obj,
    prefix=""
) {

    const output = [];

    if (
        obj &&
        typeof obj === "object"
    ) {

        for (
            const [key,value]
            of Object.entries(obj)
        ) {

            const current =
                prefix
                    ? `${prefix}.${key}`
                    : key;

            if (
                value &&
                typeof value === "object"
            ) {

                output.push(
                    ...flattenJson(
                        value,
                        current
                    )
                );

            }

            else {

                output.push({

                    path:
                        current,
                    value,
                    source:
                        "embedded_metadata"

                });

            }

        }

    }

    return output;

}
// ------------------------------------------------------------
// Encoding normalization
// ------------------------------------------------------------

function normalizeEncoding(value) {

    if (!value) {

        return null;

    }

    const v =
        String(value)
            .toLowerCase();

    if (
        v.includes("float32")
    ) {

        return "float32";

    }

    if (
        v.includes("int16")
    ) {

        return "int16";

    }

    if (
        v.includes("float64")
    ) {

        return "float64";

    }

    return v;

}

// ------------------------------------------------------------
// Reconstruction instruction builder
// ------------------------------------------------------------

function buildReconstructionInstructions(
    metadataFields,
    inferredCandidate = null
) {

    const lookup = {};

    for (const field of metadataFields) {

        lookup[
            field.path.toLowerCase()
        ] =
            field.value;

    }
    function find(names) {

        for (
            const key of Object.keys(lookup)
        ) {

            for (
                const name of names
            ) {

                if (
                    key.endsWith(
                        name.toLowerCase()
                    )
                ) {

                    return lookup[key];

                }

            }

        }

        return null;

    }
    const instructions = {

        encoding:
            normalizeEncoding(
                find([
                    "waveformencoding",
                    "waveformtype",
                    "encoding"
                ])
            ),
        endianness:
            find([
                "endianness"
            ]),
        bytes_per_sample:
            find([
                "waveformbytesperelement",
                "bytespersample"
            ]),
        samples:
            find([
                "waveformlength",
                "samples"
            ]),
        offset:
            inferredCandidate?.offset ?? 0,
        scale:
            find([
                "scale"
            ]),
        offset_value:
            find([
                "offset_value",
                "offsetvalue"
            ])

    };
    if (inferredCandidate) {

        if (!instructions.encoding) {

            instructions.encoding =
                inferredCandidate.encoding;

        }
        if (!instructions.endianness) {

            instructions.endianness =
                inferredCandidate.endianness;

        }
        if (!instructions.samples) {

            instructions.samples =
                inferredCandidate.length;

        }
        instructions.offset =
            inferredCandidate.offset ?? 0;

    }
    if (!instructions.encoding) {

        instructions.encoding =
            "float32";

    }
    if (!instructions.endianness) {

        instructions.endianness =
            "little";

    }
    if (!instructions.bytes_per_sample) {

        if (
            instructions.encoding === "float32"
        ) {

            instructions.bytes_per_sample = 4;

        }

        else if (
            instructions.encoding === "float64"
        ) {

            instructions.bytes_per_sample = 8;

        }

        else if (
            instructions.encoding === "int16"
        ) {

            instructions.bytes_per_sample = 2;

        }

    }
    if (
        instructions.scale === null ||
        instructions.scale === undefined
    ) {

        instructions.scale = 1.0;

    }
    if (
        instructions.offset_value === null ||
        instructions.offset_value === undefined
    ) {

        instructions.offset_value = 0.0;

    }
    return instructions;

}
// ------------------------------------------------------------
// Parse embedded metadata
// ------------------------------------------------------------

function parseEmbeddedMetadata(data) {

    const metadata =
        extractJsonRegion(data);

    const jsonText =
    JSON.stringify(metadata);

    const jsonOffset =
        data.indexOf(
            Buffer.from(jsonText)
        );
    const result = {
        metadata_fields: [],
        channel_labels: [],
        reconstruction_instructions_source: null
    };

    if (!metadata) {

        return result;

    }
    const fields =
        flattenJson(
            metadata
        );
    result.metadata_fields =
        fields;
    const channelLabels = [];

    for (const field of fields) {
        if (
            field.path === "metadata.channelLabels.0" ||
            field.path.startsWith("metadata.channelLabels.")
        ) {
            const match =
                field.path.match(/^metadata\.channelLabels\.(\d+)$/);

            if (
                match &&
                typeof field.value === "string" &&
                field.value.trim()
            ) {
                channelLabels[Number(match[1])] =
                    field.value.trim();
            }
        }
    }

    result.channel_labels = channelLabels.filter(
        label => typeof label === "string" && label.length > 0
    );

    let waveformLength = null;

    let waveformByteLength = null;

    let waveformEncoding = null;
    for (const field of fields) {

        if (
            field.path.endsWith(
                "waveformLength"
            )
        ) {

            waveformLength =
                field.value;

        }
        if (
            field.path.endsWith(
                "waveformByteLength"
            )
        ) {

            waveformByteLength =
                field.value;

        }
        if (
            field.path.endsWith(
                "waveformEncoding"
            )
        ) {

            waveformEncoding =
                field.value;

        }

    }
    if (
        waveformLength &&
        waveformByteLength
    ) {
        result.reconstruction_instructions_source =
        {
            samples:
                waveformLength,
            byte_length:
                waveformByteLength,
            encoding:
                waveformEncoding ||
                "Float32Array",
            offset:
                jsonOffset + jsonText.length,
        };
    }

    return result;
}
// ------------------------------------------------------------
// Float32 statistics
// ------------------------------------------------------------

function floatStats(
    data,
    offset
) {

    const values = [];

    const count =
        Math.floor(
            (data.length - offset) / 4
        );
    if (count < 64) {

        return null;

    }
    for (
        let i = 0;
        i < count;
        i++
    ) {

        let value;

        try {

            value =
                data.readFloatLE(
                    offset + i * 4
                );

        }

        catch(e) {

            break;

        }
        if (
            Number.isFinite(value)
        ) {

            values.push(value);

        }

    }
    if (
        values.length < 64
    ) {

        return null;

    }
    let min =
        values[0];

    let max =
        values[0];

    let sum = 0;
    for (const value of values) {

        if (value < min) {

            min = value;

        }
        if (value > max) {

            max = value;

        }
        sum += value;

    }
    return {

        count:
            values.length,
        min,
        max,
        mean:
            sum / values.length

    };

}

// ------------------------------------------------------------
// Binary waveform hypothesis mapper
// ------------------------------------------------------------

function findWaveformCandidate(data) {

    let best = null;
    for (
        let offset = 0;
        offset < Math.min(
            data.length,
            8192
        );
        offset += 4
    ) {

        const stats =
            floatStats(
                data,
                offset
            );
        if (!stats) {

            continue;

        }
        let confidence =
            0.5;
        if (
            stats.min > -10 &&
            stats.max < 10
        ) {

            confidence += 0.4;

        }
        if (
            Math.abs(stats.mean) < 1
        ) {

            confidence += 0.099;

        }
        const candidate =
        {

            type:
                "waveform_candidate",
            offset,
            encoding:
                "float32",
            endianness:
                "little",
            length:
                stats.count,
            statistics:
                stats,
            confidence:
                Math.min(
                    confidence,
                    0.999
                )

        };
        if (
            !best ||
            candidate.confidence >
            best.confidence
        ) {

            best = candidate;

        }

    }
    return best;

}
// ------------------------------------------------------------
// Generate IR ingestion hypothesis
// ------------------------------------------------------------

function generateHypothesis(data, filename = "binary") {

    const result =
    {

        file:
            path.basename(
                filename
            ),
        size_bytes:
            data.length,
        metadata_fields:
            [],
        arrays:
            [],
        reconstruction_instructions:
            null,
        decision:
            {}

    };
    const metadataDetection =
        detectEmbeddedMetadata(
            data
        );
    // --------------------------------------------------------
    // Path 1:
    // Embedded metadata reconstruction
    // --------------------------------------------------------

    if (
        metadataDetection.detected
    ) {

        const parsed =
            parseEmbeddedMetadata(data);
        result.metadata_fields =
            parsed.metadata_fields;
        result.channel_labels =
            parsed.channel_labels;
        result.reconstruction_instructions =
            buildReconstructionInstructions(
                result.metadata_fields,
                parsed.reconstruction_instructions_source
            );
        if (
            !result.reconstruction_instructions.samples ||
            !result.reconstruction_instructions.bytes_per_sample ||
            !result.reconstruction_instructions.encoding
        ) {
            result.decision = {
                can_create_ir: false,
                confidence: 0,
                mode: "incomplete_reconstruction_instructions",
                error: "Embedded metadata found but waveform reconstruction information is incomplete"
            };

            return result;
        }

result.decision =
{
    can_create_ir: true,
            confidence:
                1.0,
            mode:
                "embedded_reconstruction_instructions"

        };
        return result;

    }

    // --------------------------------------------------------
    // Path 2:
    // Heuristic waveform discovery
    // --------------------------------------------------------

    const candidate =
        findWaveformCandidate(
            data
        );
    if (candidate) {

        result.arrays =
        [

            candidate

        ];
        result.reconstruction_instructions =
            buildReconstructionInstructions(
                [],
                candidate
            );
        result.decision =
        {

            can_create_ir:
                true,
            confidence:
                candidate.confidence,
            mode:
                "heuristic_reconstruction_instructions"

        };
        return result;

    }

    // --------------------------------------------------------
    // Path 3:
    // Unknown binary
    // --------------------------------------------------------

    result.decision =
    {
        can_create_ir:
            false,
        confidence:
            0,
        mode:
            "unknown_binary_layout",
        error:
            "Unable to determine waveform reconstruction instructions"

    };
    return result;
}
// ------------------------------------------------------------
// Export
// ------------------------------------------------------------

module.exports = {
    generateHypothesis
};